File: sd_python/src/sd_python/utils/binning.py
from typing import Sequence

from sd_python.core.droplet import Droplet


def bin_droplets(
    droplets: Sequence[Droplet], bin_edges: Sequence[float]
) -> list[float]:
    """Bin droplets into bins defined by `bin_edges`.

    Hard-coded to sort `droplets` by the `radius` attribute in each Droplet, and to
    tally up the total mass of droplets in each bin.

    Args:
        droplets: Sequence of droplets to bin.
        bin_edges: Sequence of bin edges.

    Returns:
        List of bin values.
    """

    # Compute a list of the indices corresponding to the sorted droplets
    sorted_indices = sorted(range(len(droplets)), key=lambda i: droplets[i].radius)

    bin_values = [0.0] * (len(bin_edges) - 1)
    bin_idx = 0
    droplet_idx = 0
    for left, right in zip(bin_edges[:-1], bin_edges[1:]):
        bin_value = 0.0
        bin_r_max = right * 1e-6

        for idx in sorted_indices[droplet_idx:]:
            droplet = droplets[idx]
            if droplet.radius < bin_r_max:
                bin_value += droplet.mass * droplet.multi
                droplet_idx += 1
            else:
                break

        bin_values[bin_idx] = bin_value
        bin_idx += 1

    return bin_values

File: sd_julia-v2/src/utils/binning.jl
"""
Droplet binning for analysis and output.

Bins droplets by size for visualization and analysis purposes.
"""

"""
    bin_droplets(droplets::Vector{Droplet}, bin_edges::Vector{Float64}) -> Vector{Float64}

Bin droplets by size and compute total water mass in each bin.

Uses logarithmically-spaced bins and sorts droplets by size for
efficient binning.

# Arguments
- `droplets`: Vector of droplets to bin
- `bin_edges`: Bin edges in microns (μm)

# Returns
- Vector of total water mass (kg) in each bin

# Example
```julia
# Create logarithmically-spaced bins from 1 to 1000 microns
bin_edges = [10.0^exp for exp in range(0.0, 3.0, length=101)]
bin_values = bin_droplets(droplets, bin_edges)
```
"""
function bin_droplets(droplets::Vector{Droplet}, bin_edges::Vector{Float64})::Vector{Float64}
    n_bins = length(bin_edges) - 1
    bin_values = zeros(Float64, n_bins)
    
    # Sort indices by radius (not rcubed) for correct binning
    sorted_indices = sortperm([d.radius for d in droplets])
    
    droplet_idx = 1
    n_droplets = length(droplets)
    
    # Iterate over bin edges
    for bin_idx in 1:n_bins
        bin_value = 0.0
        r_max = bin_edges[bin_idx + 1] * 1.0e-6  # Convert microns to meters
        
        # Add all droplets that fall in this bin
        while droplet_idx <= n_droplets
            idx = sorted_indices[droplet_idx]
            droplet = droplets[idx]
            
            if droplet.radius < r_max
                bin_value += droplet.mass * Float64(droplet.multi)
                droplet_idx += 1
            else
                break
            end
        end
        
        bin_values[bin_idx] = bin_value
    end
    
    return bin_values
end
